Objects/descrobject.c

```diff
@@ -1790,3 +1790,423 @@ PyTypeObject PyProperty_Type = {
+            return NULL;
+        Py_DECREF(eq);
+        if (eq == Py_True) {
+            Py_RETURN_FALSE;
+        }
+        else {
+            Py_RETURN_TRUE;
+        }
+    }
+
+    gaobject *aa = (gaobject *)a;
+    gaobject *bb = (gaobject *)b;
+    PyObject *eq = PyObject_RichCompare(aa->origin, bb->origin, Py_EQ);
+    if (eq == NULL)
+        return NULL;
+    if (eq == Py_False) {
+        return eq;
+    }
+    Py_DECREF(eq);
+    return PyObject_RichCompare(aa->args, bb->args, Py_EQ);
+}
+
+static PyObject *
+ga_mro_entries(PyObject *self, PyObject *args)
+{
+    gaobject *alias = (gaobject *)self;
+    return PyTuple_Pack(1, alias->origin);
+}
+
+static PyObject *
+ga_instancecheck(PyObject *self, PyObject *Py_UNUSED(ignored))
+{
+    return PyErr_Format(PyExc_TypeError,
+                        "TypeError: Subscripted generics cannot be used with class and instance checks",
+                        self);
+}
+
+static PyObject *
+ga_subclasscheck(PyObject *self, PyObject *Py_UNUSED(ignored))
+{
+    return PyErr_Format(PyExc_TypeError,
+                        "TypeError: Subscripted generics cannot be used with class and instance checks",
+                        self);
+}
+
+static PyObject *
+ga_reduce(PyObject *self, PyObject *Py_UNUSED(ignored))
+{
+    gaobject *alias = (gaobject *)self;
+    return Py_BuildValue("O(OO)", Py_TYPE(alias),
+                         alias->origin, alias->args);
+}
+
+static PyObject *
+ga_setstate(PyObject *self, PyObject *state)
+{
+    gaobject *alias = (gaobject *)self;
+    PyObject *parameters = make_parameters(alias->args);
+    Py_INCREF(parameters);
+    alias->parameters = parameters;
+    Py_RETURN_NONE;
+}
+
+static PyMethodDef ga_methods[] = {
+    {"__mro_entries__", ga_mro_entries, METH_O},
+    {"__instancecheck__", ga_instancecheck, METH_O},
+    {"__subclasscheck__", ga_subclasscheck, METH_O},
+    {"__reduce__", ga_reduce, METH_NOARGS},
+    {"__setstate__", ga_setstate, METH_O},
+    {0}
+};
+
+static PyMemberDef ga_members[] = {
+    {"__origin__", T_OBJECT, offsetof(gaobject, origin), READONLY},
+    {"__args__", T_OBJECT, offsetof(gaobject, args), READONLY},
+    {"__parameters__", T_OBJECT, offsetof(gaobject, parameters), READONLY},
+    {0}
+};
+
+static PyObject *
+ga_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
+{
+    if (kwds != NULL && PyDict_GET_SIZE(kwds) != 0) {
+        return PyErr_Format(PyExc_TypeError, "GenericAlias does not support keyword arguments");
+    }
+    if (PyTuple_Size(args) != 2) {
+        return PyErr_Format(PyExc_TypeError, "GenericAlias expects 2 positional arguments");
+    }
+    PyObject *origin = PyTuple_GET_ITEM(args, 0);
+    PyObject *arguments = PyTuple_GET_ITEM(args, 1);
+    return Py_GenericAlias(origin, arguments);
+}
+
+// TODO:
+// - argument clinic?
+// - __doc__?
+// - cache?
+PyTypeObject Py_GenericAliasType = {
+    PyVarObject_HEAD_INIT(&PyType_Type, 0)
+    .tp_name = "types.GenericAlias",
```

Comment:
I know it's late for this bikeshedding but I was always a bit puzzled by the name "GenericAlias". What is it aliasing?

Especially now that it lives in `types`, wouldn't "ParametrizedType" work better?

Reply:
I think "ParametrizedType" is both long and easy to misspell (I can already tell I will write "ParameterizedType" and be confused).

Things in `type` generally are `NameType`, like `FunctionType` or `CodeType`, so perhaps `GenericType`? I agree the word "alias" is confusing.

Reply:
Agreed that GenericAlias is a lame name for the concept. It's easy to change now, but will be impossible to change once the PEP's been accepted, so it's appropriate to bikeshed about it (here, and maybe in the list). At first blush I do like GenericType.

Reply:
I brought this up on python-dev and it looks like GenericAlias is here to stay.